src/rust/engine/src/nodes.rs

```diff
@@ -301,34 +301,24 @@ impl MultiPlatformExecuteProcess {
   }
 
   pub fn lift(value: &Value) -> Result<MultiPlatformExecuteProcess, String> {
-    let constraint_parts = externs::project_multi_strs(&value, "platform_constraints");
-    if constraint_parts.len() % 2 != 0 {
-      return Err("Error parsing platform_constraints: odd number of parts".to_owned());
-    }
-    let constraint_key_pairs: Vec<_> = constraint_parts
-      .chunks_exact(2)
-      .map(|constraint_key_pair| {
-        (
-          PlatformConstraint::try_from(&constraint_key_pair[0]).unwrap(),
-          PlatformConstraint::try_from(&constraint_key_pair[1]).unwrap(),
-        )
-      })
-      .collect();
+    let constraints = externs::project_multi_strs(&value, "platform_constraints")
+      .into_iter()
+      .map(|constraint| PlatformConstraint::try_from(&constraint))
+      .collect::<Result<Vec<_>, _>>()?;
     let processes = externs::project_multi(&value, "processes");
-    if constraint_parts.len() / 2 != processes.len() {
+    if constraints.len() != processes.len() {
       return Err(format!(
         "Sizes of constraint keys and processes do not match: {} vs. {}",
-        constraint_parts.len() / 2,
+        constraints.len(),
         processes.len()
       ));
     }
 
-    let mut request_by_constraint: BTreeMap<(PlatformConstraint, PlatformConstraint), Process> =
-      BTreeMap::new();
-    for (constraint_key, execute_process) in constraint_key_pairs.iter().zip(processes.iter()) {
+    let mut request_by_constraint: BTreeMap<PlatformConstraint, Process> = BTreeMap::new();
+    for (constraint, execute_process) in constraints.into_iter().zip(processes.iter()) {
       let underlying_req =
-        MultiPlatformExecuteProcess::lift_execute_process(execute_process, constraint_key.1)?;
-      request_by_constraint.insert(constraint_key.clone(), underlying_req.clone());
+        MultiPlatformExecuteProcess::lift_execute_process(execute_process, constraint)?;
```

Comment:
I don't think it works to pass the execution platform here in place of the target_platform. In fact it might screw up some of the work greg did around adding the target platform to the ProcessResult. May need to remove the field from MPEP entirely? But that also has implications for caching the results of the request in the lmdb. Currently we build a unique key from the requests present in the MPEP, but if the target platform isn't explicit anymore, the same request could have different output depending on the target platform that is injected into the subgraph by the requester.

Reply:
I see what you're saying, and I have patches coming to address that. But I think that the "Platform::current" constraint that we put on all pex executions (here: https://github.com/pantsbuild/pants/blob/77f68ce5521dd9acae951c223cf71775c0b0c44f/src/python/pants/backend/python/rules/pex.py#L362) prevents it from being an actual issue currently.

I think that I can reorder the patches to remove the necessity of that constraint by consuming the platform param. But I also think that while that constraint is in place 
this change is safe.

Reply:
Ah yeah I see what you mean about pex. We were already setting target_platform==execution_platform. So it should be safe, but in general when crossbuilding it would be unsafe to cache without target platform information recorded somehow. If you have a good handle on follow ups that make that work safely then I am okay with this.

Reply:
For my edification can the Params in use by an intrinsic rule be added to its cache key?